### ISSUE TITLE
HH:mm accept single character minutes value

### ISSUE THREAD

**Issue by bohdanbirdie** [date-fns/date-fns#2131]
According to the documentation, the `mm` pattern should not accept a single character value (https://date-fns.org/v2.16.1/docs/parse)
![image](https://user-images.githubusercontent.com/23266928/103092178-14a0ed80-45ff-11eb-99ff-14e238fe15ae.png)

But, I'm able to parse the string regardless of the numbers of characters.
Is this a but or I misunderstood the documation?
My goal is to validate the time input, but it must specifically be 4 characters 

https://codesandbox.io/s/date-fns-v2-forked-myt1d?file=/src/index.js:0-149
```TS
import { parse } from "date-fns";

const shouldNotParse = parse("20:3", "HH:mm", new Date()); // Thu Dec 24 2020 20:03:00 GMT+0200 (Eastern European Standard Time)
const shoudParse = parse("20:30", "HH:mm", new Date()); // Thu Dec 24 2020 20:30:00 GMT+0200 (Eastern European Standard Time)
```

**Comment by simlu** [comment]
Ran into this one

```
const { parse } = require("date-fns");

const r = parse(`19701 12345Z`, 'yyyyD HmmssX', new Date(), { useAdditionalDayOfYearTokens: true });
console.log(r);
// => 1970-01-01T12:34:05.000Z
```

The expected output should be `1970-01-01T01:23:45Z`

**Comment by fturmel** [comment]
It's been acknowledged that there are strictness issues with `parse`, and it appears to be on the roadmap for v3 #2561

@simlu specifically regarding your example, the `H` token can match 0-23, so it's ambiguous whether the hour should be 1 or 12. If your input string was `19701 24345Z` it would work as you expect because `24` isn't a proper value for `H` so the parser would only consume one character for that token.

Considering how `parse` works, I can think of two ways to get rid of the ambiguity:

1) add a separator after `H`

```js
parse(`19701 1:2345Z`, "yyyyD H:mmssX", new Date(), {
  useAdditionalDayOfYearTokens: true
}); // 1970-01-01T01:23:45.000Z
```

2) use `HH` and leading zeroes so that there are always precisely 2 characters being read for the hour

```js
parse(`19701 012345Z`, "yyyyD HHmmssX", new Date(), {
  useAdditionalDayOfYearTokens: true
}); // 1970-01-01T01:23:45.000Z
```

Hope this helps!

**Comment by simlu** [comment]
@fturmel Thank you for the quick reply! Glad to hear it's in the road map. 

Padding the hours with zeros is exactly how I ended up solving his.

However I disagree that there is disambiguity here. `mm` and `ss` match exactly two characters according to the documentation. If that was actually enforced, things would always parse correctly.

**Comment by fturmel** [comment]
> However I disagree that there is disambiguity here. mm and ss match exactly two characters according to the documentation. If that was actually enforced, things would always parse correctly.

Yeah, I did consider that... It's not so much token strictness that comes into play here (`12` _is_ valid for `H`), but the fact that the token values are extracted left to right. There would need to be a priority system based on token specificity instead.

Are there enough of these edge-case scenarios out there that it's worth the added perf cost and complexity? Maybe it's simpler than I'm imagining, and just doing a first pass for tokens that consume fixed-length values would be enough.

Interesting to consider. Maybe others can chime-in on this too.

**Comment by simlu** [comment]
@fturmel Regex solve this nicely. But I'm not sure how optimised this library is and adding a parser might be a significant performance overhead?

**Comment by alan-nf** [comment]
This is basically the same problem, right:

```
> require('date-fns').isMatch("3:01", "HH:mm")
true
```
